src/renderers/shared/fiber/ReactFiberCompleteWork.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberCompleteWork
+ * @flow
+ */
+
+'use strict';
+
+import type { ReactCoroutine, ReactYield } from 'ReactCoroutine';
+import type { Fiber } from 'ReactFiber';
+
+var ReactChildFiber = require('ReactChildFiber');
+var ReactTypesOfWork = require('ReactTypesOfWork');
+var {
+  IndeterminateComponent,
+  FunctionalComponent,
+  ClassComponent,
+  HostComponent,
+  CoroutineComponent,
+  YieldComponent,
+} = ReactTypesOfWork;
+
+function transferOutput(unitOfWork : Fiber) {
+  var child = unitOfWork.child;
+  if (child && !child.sibling) {
+    // Common case
+    unitOfWork.output = child.output;
+    return;
+  }
+  var output = [];
+  while (child) {
+    output.push(child.output);
+    child = child.sibling;
+  }
+  unitOfWork.output = output;
+}
+
+function recursivelyFillYields(yields, output) {
+  if (Array.isArray(output)) {
+    for (var i = 0; i < output.length; i++) {
+      recursivelyFillYields(yields, output[i]);
+    }
+  } else {
+    yields.push(output);
+  }
+}
+
+function handleCoroutine(unitOfWork : Fiber) {
+  var coroutine = (unitOfWork.input : ?ReactCoroutine);
+  if (!coroutine) {
+    throw new Error('Should be resolved by now');
+  }
+
+  // Hacky McHack
+  if (!unitOfWork.hasPendingChanges) {
```

Comment:
Yea, this needs to change. It's terrible. That's what I was thinking about and wanted to discuss yesterday. I wanted to avoid overloading the structure too much but also avoid too many unnecessary nodes.